### PR TITLE
Feature 314/checkin notes non blocking #314

### PR DESCRIPTION
**Pull request by Maroju100**
Convert checkin notes to non blocking

### CODE REVIEW

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java [hunk 1]
```diff
@@ -14,92 +13,176 @@
...
+                                @Named(TaskExecutors.IO) ExecutorService ioExecutorService) {
         this.checkinNoteServices = checkinNoteServices;
+        this.eventLoopGroup = eventLoopGroup;
+        this.ioExecutorService = ioExecutorService;
     }
 
-
     @Error(exception = CheckinNotesBadArgException.class)
     public HttpResponse<?> handleBadArgs(HttpRequest<?> request, CheckinNotesBadArgException e) {
+        LOG.info("Throwing the wrong exception");
```
Comment (bhatiam12) [inline server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java:55]: were these loggers added for testing purpose ? there are many in this class so please remove it that are unnecessary.

Comment (Maroju100) [reply server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java]: Updated, removed the unnecessary logging, thanks.

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java [hunk 2]
```diff
@@ -45,61 +55,119 @@ public CheckinNoteController(CheckinNoteServices checkinNoteServices) {
...
-        return HttpResponse.created(newCheckinNote)
-                .headers(headers -> headers.location(URI.create(String.format("%s/%s", request.getPath(), newCheckinNote.getId()))));
-
+    public Single<HttpResponse<CheckinNote>> createCheckinNote(@Body @Valid CheckinNoteCreateDTO checkinNote,
+                                                                    HttpRequest<CheckinNoteCreateDTO> request) {
+        return Single.fromCallable(() -> checkinNoteServices.save(new CheckinNote(checkinNote.getCheckinid(),
+                checkinNote.getCreatedbyid(), checkinNote.getDescription())))
+                .observeOn(Schedulers.from(eventLoopGroup))
+                .map(createdCheckinNote -> {
+                    //Using code block rather than lambda so we can log what thread we're in
```
Comment (meyerinj) [inline server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java:103]: This comment was here for demonstrative purposes. You can get rid of it and change this to a lambda and the result will be a much clearer read.

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java [hunk 3]
```diff
@@ -45,61 +55,119 @@ public CheckinNoteController(CheckinNoteServices checkinNoteServices) {
...
-                URI.create(String.format("%s/%s", request.getPath(), updateCheckinNote.getId()))))
-                .body(updateCheckinNote);
+    public Single<HttpResponse<CheckinNote>> updateCheckinNote(@Body @Valid CheckinNote checkinNote,
+                                            HttpRequest<CheckinNote> request) {
+        if (checkinNote == null) {
+            return Single.just(HttpResponse.ok());
+        }
+        return Single.fromCallable(() -> checkinNoteServices.update(checkinNote))
+            .observeOn(Schedulers.from(eventLoopGroup))
+            .map(updatedCheckinNote -> //This lambda expression is the preferred way to do this kind of simple mapping.
```
Comment (meyerinj) [inline server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java:126]: The lambda will look a lot like this. This comment was also only there to be demonstrative.

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteServicesImpl.java [hunk 4]
```diff
@@ -48,6 +46,7 @@ public CheckinNote save(CheckinNote checkinNote) {
         Boolean isAdmin = securityService != null ? securityService.hasRole(RoleType.Constants.ADMIN_ROLE) : false;
 
         if (checkinNote != null) {
+            LOG.info("Making multiple calls within this I/O thread as thread switching is costly.");
```
Comment (meyerinj) [inline server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteServicesImpl.java:49]: These logs were important in the draft PR to demonstrate that the threading worked but we don't need them on develop

**Review by meyerinj — CHANGES_REQUESTED** [review]

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteServicesImpl.java (hunk 4 above)
Comment (Maroju100) [reply server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteServicesImpl.java]: removed the LOG.

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java (hunk 2 above)
Comment (Maroju100) [reply server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java]: Updated

File: server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java (hunk 3 above)
Comment (Maroju100) [reply server/src/main/java/com/objectcomputing/checkins/services/checkin_notes/CheckinNoteController.java]: Updated

**Review by meyerinj — APPROVED** [review]

**Review by mkimberlin — APPROVED** [review]

**Pull request closed without merging**